src/cpptoolsconfigprovider.ts

```diff
@@ -0,0 +1,120 @@
+import path = require("path");
+import * as vscode from "vscode";
+import * as cpptools from "vscode-cpptools";
+import { getMesonBuildOptions, getMesonCompilers, getMesonDependencies } from "./introspection";
+var fs = require("fs");
+
+export class CpptoolsProvider implements cpptools.CustomConfigurationProvider {
+  cppToolsAPI?: cpptools.CppToolsApi;
+  private buildDir: string = "";
+
+  constructor(buildDir: string) {
+    this.buildDir = buildDir;
+  }
+
+  name = "Meson build";
+  extensionId = "mesonbuild.mesonbuild";
+
+  canProvideBrowseConfiguration(token?: vscode.CancellationToken | undefined): Thenable<boolean> {
+    return new Promise<boolean>((resolve) => {
+      if (this.buildDir !== "") {
+        resolve(true);
+      } else {
+        // Wait for this.buildDir to not be ""
+        const interval = setInterval(() => {
+          if (this.buildDir !== "") {
+            clearInterval(interval);
+            this.refresh(this.buildDir);
+            resolve(true);
+          }
+        }, 100);
+      }
+    });
+  }
+
+  async provideBrowseConfiguration(
+    token?: vscode.CancellationToken | undefined,
+  ): Promise<cpptools.WorkspaceBrowseConfiguration | null> {
+    let browseConfig: cpptools.WorkspaceBrowseConfiguration = {
+      browsePath: [],
+      compilerPath: "",
+      compilerArgs: [],
+      standard: "c++17",
+    };
+
+    const dependencies = await getMesonDependencies(this.buildDir);
+    for (let dep of dependencies) {
+      if (dep.compile_args) {
+        for (let arg of dep.compile_args) {
+          if (arg.startsWith("-I")) {
+            browseConfig.browsePath.push(arg.slice(2));
+          }
+        }
+      }
+    }
+
+    let machine = "";
+    const buildOptions = await getMesonBuildOptions(this.buildDir);
+    for (let option of buildOptions) {
+      if (option.name === "cpp_std") {
+        browseConfig = Object.assign({}, browseConfig, { standard: option.value });
+        machine = option.machine;
+      }
+    }
```

Comment:
Is it possible for  us to not have a machine?

Reply:
I tested the simple [meson test-case gtk-doc](https://github.com/mesonbuild/meson/tree/master/test%20cases/frameworks/10%20gtk-doc) with a native build. A machine was also defined ("host"). So it looks like meson always defines a machine, even with empty/default options.

Reply:
Let's add an assert and make `machine` `string | undefined`

Reply:
Alright! Added a commit to handle an undefined machine. I used an if rather than assert because we are still able to provide a browsePath without this option.